File: casatables/syscall_tracer.py
#!/usr/bin/env python3
"""
Python version of the syscall tracer using python-casacore.
This creates real casacore table operations for accurate syscall comparison.
"""

import numpy as np
import tempfile
import os
import json
import sys

def check_casacore():
    """Check if python-casacore is available"""
    try:
        import casacore.tables as ct
        return True
    except ImportError:
        print("python-casacore not found. Install with: pip install python-casacore")
        return False

def main():
    if not check_casacore():
        sys.exit(1)

    import casacore.tables as ct

    print("Starting Python syscall tracer with casacore...")

    try:
        # Simple operations for syscall analysis
        n_rows = 100
        data_shape = [32, 4]

        with tempfile.TemporaryDirectory() as tmp_dir:
            table_path = os.path.join(tmp_dir, "syscall_test_python.ms")

            print("Creating casacore table...")

            # Create table description
            desc = ct.maketabdesc([
                ct.makescacoldesc("TIME", 0.0, comment="Observation time"),
                ct.makescacoldesc("ANTENNA1", 0, comment="First antenna"),
                ct.makescacoldesc("ANTENNA2", 0, comment="Second antenna"),
                ct.makescacoldesc("FLAG_ROW", False, comment="Row flag"),
                ct.makearrcoldesc("DATA", np.complex64(0), shape=data_shape, comment="Visibility data"),
                ct.makearrcoldesc("FLAG", False, shape=data_shape, comment="Data flags")
            ])

            # Create the table
            table = ct.table(table_path, desc, nrow=n_rows, readonly=False)

            print("Writing data...")

            # Prepare test data
            data_matrix = np.zeros(data_shape, dtype=np.complex64)
            flag_matrix = np.zeros(data_shape, dtype=bool)

            # Write a few rows
            for row_idx in range(min(10, n_rows)):  # Write fewer rows for efficiency
                # Create test data with some pattern
                for i in range(data_shape[0]):
                    for j in range(data_shape[1]):
                        idx = i * data_shape[1] + j
                        data_matrix[i, j] = complex(float(idx % 100), 0.0)
                        flag_matrix[i, j] = (idx + row_idx) % 13 == 0

                # Write to table
                table.putcell("TIME", row_idx, float(row_idx))
                table.putcell("ANTENNA1", row_idx, (row_idx % 128))
                table.putcell("ANTENNA2", row_idx, ((row_idx + 1) % 128))
                table.putcell("FLAG_ROW", row_idx, (row_idx % 2 == 0))
                table.putcell("DATA", row_idx, data_matrix)
                table.putcell("FLAG", row_idx, flag_matrix)

            print("Reading data...")

            # Read some data back
            for row_idx in range(min(5, n_rows)):
                time_val = table.getcell("TIME", row_idx)
                ant1_val = table.getcell("ANTENNA1", row_idx)
                ant2_val = table.getcell("ANTENNA2", row_idx)
                flag_row_val = table.getcell("FLAG_ROW", row_idx)
                data_matrix = table.getcell("DATA", row_idx)
                flag_matrix = table.getcell("FLAG", row_idx)

                # Some processing to generate syscalls
                processed_time = time_val * 2.0
                baseline = ant1_val + ant2_val
                combined_flag = flag_row_val or np.any(flag_matrix)

            print("Performing additional I/O operations...")

            # Additional file I/O operations
            test_file = os.path.join(tmp_dir, "test_data.npy")
            np.save(test_file, data_matrix)
            loaded_data = np.load(test_file)

            # JSON operations
            metadata = {
                "n_rows": n_rows,
                "data_shape": data_shape,
                "test_value": 42,
                "library": "python-casacore"
            }
            metadata_file = os.path.join(tmp_dir, "metadata.json")
            with open(metadata_file, 'w') as f:
                json.dump(metadata, f)

            with open(metadata_file, 'r') as f:
                loaded_metadata = json.load(f)

            # Close table
            table.close()

        print("Python syscall tracer with casacore completed successfully.")

    except Exception as e:
        print(f"Error in Python syscall tracer: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == "__main__":
    main()

File: casatables/syscall_tracer.cpp
#include <iostream>
#include <vector>
#include <complex>
#include <fstream>
#include <filesystem>
#include <cstdlib>
#include <casacore/casa/aips.h>
#include <casacore/tables/Tables/Table.h>
#include <casacore/tables/Tables/TableDesc.h>
#include <casacore/tables/Tables/SetupNewTab.h>
#include <casacore/tables/Tables/ScalarColumn.h>
#include <casacore/tables/Tables/ArrayColumn.h>
#include <casacore/tables/Tables/ScaColDesc.h>
#include <casacore/tables/Tables/ArrColDesc.h>
#include <casacore/casa/Arrays/Vector.h>
#include <casacore/casa/Arrays/Matrix.h>
#include <casacore/casa/Arrays/IPosition.h>
#include <casacore/casa/BasicSL/Complex.h>

// Use the casacore namespace
using namespace casacore;

int main() {
    std::cout << "Starting C++ syscall tracer with casacore..." << std::endl;

    try {
        // CasaCore initialization is handled automatically

        // Simple operations for syscall analysis
        uInt n_rows = 100;
        IPosition data_shape(2, 32, 4);  // 32x4 matrix

        // Create temporary directory
        auto tmp_dir = std::filesystem::temp_directory_path() / "syscall_test_cpp";
        std::filesystem::create_directory(tmp_dir);
        std::string table_path = (tmp_dir / "syscall_test_cpp.ms").string();

        std::cout << "Creating casacore table..." << std::endl;

        // Create table description
        TableDesc td("syscall_test", "1", TableDesc::Scratch);

        // Add scalar columns
        td.addColumn(ScalarColumnDesc<Double>("TIME", "Observation time"));
        td.addColumn(ScalarColumnDesc<Int>("ANTENNA1", "First antenna"));
        td.addColumn(ScalarColumnDesc<Int>("ANTENNA2", "Second antenna"));
        td.addColumn(ScalarColumnDesc<Bool>("FLAG_ROW", "Row flag"));

        // Add array columns
        td.addColumn(ArrayColumnDesc<Complex>("DATA", "Visibility data", data_shape, ColumnDesc::FixedShape));
        td.addColumn(ArrayColumnDesc<Bool>("FLAG", "Data flags", data_shape, ColumnDesc::FixedShape));

        // Create the table
        SetupNewTable setup(table_path, td, Table::New);
        Table table(setup, n_rows);

        std::cout << "Writing data..." << std::endl;

        // Write mode via env: WRITE_MODE={column_put|column_put_bulk}
        const char *mode_env = std::getenv("WRITE_MODE");
        std::string write_mode = mode_env ? std::string(mode_env) : std::string("column_put");

        // Create column objects
        ScalarColumn<Double> time_col(table, "TIME");
        ScalarColumn<Int> ant1_col(table, "ANTENNA1");
        ScalarColumn<Int> ant2_col(table, "ANTENNA2");
        ScalarColumn<Bool> flag_row_col(table, "FLAG_ROW");
        ArrayColumn<Complex> data_col(table, "DATA");
        ArrayColumn<Bool> flag_col(table, "FLAG");

        if (write_mode == "column_put_bulk") {
            // Bulk path: write arrays for first 3 rows in one call each, mirroring Rust bulk
            const uInt rows_to_write = 3;
            IPosition arrShape(3, rows_to_write, data_shape[0], data_shape[1]);

            Array<Complex> data_arr(arrShape);
            Array<Bool> flag_arr(arrShape);

            // Fill arrays: layout is [row, i, j]
            {
                Array<Complex>::iterator it = data_arr.begin();
                for (uInt r = 0; r < rows_to_write; ++r) {
                    for (uInt i = 0; i < data_shape[0]; ++i) {
                        for (uInt j = 0; j < data_shape[1]; ++j) {
                            uInt idx = i * data_shape[1] + j;
                            *it++ = Complex(static_cast<float>(idx), 0.0f);
                        }
                    }
                }
            }
            {
                Array<Bool>::iterator it = flag_arr.begin();
                for (uInt r = 0; r < rows_to_write; ++r) {
                    for (uInt i = 0; i < data_shape[0]; ++i) {
                        for (uInt j = 0; j < data_shape[1]; ++j) {
                            uInt idx = i * data_shape[1] + j;
                            *it++ = (idx % 13 == 0);
                        }
                    }
                }
            }

            // Put arrays for columns at once
            data_col.putColumn(data_arr);
            flag_col.putColumn(flag_arr);

            // Note: to mirror Rust bulk, we skip scalar column writes here
        } else {
            // Per-row path: write a few rows
            for (uInt row_idx = 0; row_idx < 3; ++row_idx) {
                // Create test data
                Matrix<Complex> data_matrix(data_shape);
                Matrix<Bool> flag_matrix(data_shape, false);

                // Fill data matrix with test values
                for (uInt i = 0; i < data_shape[0]; ++i) {
                    for (uInt j = 0; j < data_shape[1]; ++j) {
                        uInt idx = i * data_shape[1] + j;
                        data_matrix(i, j) = Complex(static_cast<float>(idx), 0.0f);
                        flag_matrix(i, j) = (idx % 13 == 0);  // Some pattern for flags
                    }
                }

                // Write data to columns
                time_col.put(row_idx, static_cast<Double>(row_idx));
                ant1_col.put(row_idx, static_cast<Int>(row_idx % 128));
                ant2_col.put(row_idx, static_cast<Int>((row_idx + 1) % 128));
                flag_row_col.put(row_idx, (row_idx % 2 == 0));
                data_col.put(row_idx, data_matrix);
                flag_col.put(row_idx, flag_matrix);
            }
        }

        // Clean up
        table.markForDelete();
        std::filesystem::remove_all(tmp_dir);

        std::cout << "C++ syscall tracer with casacore completed successfully." << std::endl;

    } catch (const AipsError& e) {
        std::cerr << "CasaCore error: " << e.getMesg() << std::endl;
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "Standard error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
